receiver/opencensusreceiver/opencensus.go

```diff
@@ -253,54 +256,46 @@ func (ocr *Receiver) httpServer() *http.Server {
 }
 
 func (ocr *Receiver) startServer() error {
-	err := errAlreadyStarted
+	startErr := errAlreadyStarted
 	ocr.startServerOnce.Do(func() {
-		errChan := make(chan error, 1)
+		// Register the grpc-gateway on the HTTP server mux
+		c := context.Background()
+		// TODO: check use of grpc.WithInsecure in the context of
+		// https://github.com/open-telemetry/opentelemetry-service/issues/170
+		opts := []grpc.DialOption{grpc.WithInsecure()}
+		endpoint := ocr.ln.Addr().String()
+
+		err := agenttracepb.RegisterTraceServiceHandlerFromEndpoint(c, ocr.gatewayMux, endpoint, opts)
+		if err != nil {
+			startErr = err
+			return
+		}
+
+		err = agentmetricspb.RegisterMetricsServiceHandlerFromEndpoint(c, ocr.gatewayMux, endpoint, opts)
+		if err != nil {
+			startErr = err
+			return
+		}
+
+		// Start the gRPC and HTTP/JSON (grpc-gateway) servers on the same port.
+		m := cmux.New(ocr.ln)
+		grpcL := m.MatchWithWriters(
+			cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"),
+			cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc+proto"))
+
+		httpL := m.Match(cmux.Any())
+
 		go func() {
-			// Register the grpc-gateway on the HTTP server mux
-			c := context.Background()
-			opts := []grpc.DialOption{grpc.WithInsecure()}
-			endpoint := ocr.ln.Addr().String()
-
-			err := agenttracepb.RegisterTraceServiceHandlerFromEndpoint(c, ocr.gatewayMux, endpoint, opts)
-			if err != nil {
-				errChan <- err
-				return
-			}
-
-			err = agentmetricspb.RegisterMetricsServiceHandlerFromEndpoint(c, ocr.gatewayMux, endpoint, opts)
-			if err != nil {
-				errChan <- err
-				return
-			}
-
-			// Start the gRPC and HTTP/JSON (grpc-gateway) servers on the same port.
-			m := cmux.New(ocr.ln)
-			grpcL := m.MatchWithWriters(
-				cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"),
-				cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc+proto"))
-
-			httpL := m.Match(cmux.Any())
-			go func() {
-				errChan <- ocr.serverGRPC.Serve(grpcL)
-			}()
-			go func() {
-				errChan <- ocr.httpServer().Serve(httpL)
-			}()
-			errChan <- m.Serve()
+			ocr.host.ReportFatalError(ocr.serverGRPC.Serve(grpcL))
```

Comment:
Isn't it needed to check if Serve() returns nil?

Reply:
Good question. 

In practice these are all blocking calls that will only return when the server is closed (due to errors or explicit stop/close). Prior to a shutdown we want to be notified if they terminate (even if returning nil) since it means that the receiver in question is not running anymore - that said it seems a server implementation error if they return nil without corresponding stop/close call (in such case the log message won't be helpful at all).

In a shutdown triggered via ReportFatalError, ie.: "serve" call returning, only the first one is reported in the logs. For a shutdown via SIGTERM no errors reported via ReportFatalError are logged. Notice that it is implementation dependent if the "Serve" will return nil or an error in case the corresponding stop/close is called: e.g.: http always returns error while grpc can return nil if stop/close is called.